### PR TITLE
Enhanced Exception Handling and Other Minor Fixes

### PR DESCRIPTION
**Pull request by asdfghjkxd**
A few fixes are implemented in this PR: Exception Handling, Proposed Improvements to Functional Programming Monads and Other Minor Fixes to Code Quality.

**Exception Handling**

Most prominently, the exception handling system has been enhanced. More exception classes are created to break down previously confusing exceptions into more comprehensible ones that tell you what went wrong, rather than just a catch-all exception category which obscures the true reason why an error has occurred.

**Proposed Improvements to Functional Programming Monads**

After observing some of the functional programming patterns which were used in the code, it has become clear that a functional programming interface which allows for inner functions to throw checked exceptions that would be floated back to the caller is necessary. There is a significant number of patterns whereby `Optional` was used in a manner that would not require the use of `Optional` in the first place.

In its place, I have proposed 2 new monads, `ExOptional` and `ExFunction`, which are the "exceptional" successor to the `Optional` monad. Do note that only the methods which I have observed being in use by the existing code base were implemented; more methods could be implemented in the future should the team see fit.

`ExOptional` mimics the behaviour of `Optional`, but it permits the use of functions that throw checked exceptions. These exceptions are floated back to the caller of the `ExOptional` methods. So far, only the `map` and `flatMap` methods support this exception floating behaviour.

`ExFunction` mimics the behaviour of the functional interface `Function`, by allowing users to declare functions that throw exceptions.

For example, the `ExOptional` class could be used to simplify this chunk of code:

```java
public Optional<ReadOnlyTaskWise> readTaskWise(Path filePath) throws FileStorageLoadException {
    requireNonNull(filePath);

    Optional<JsonSerializableTaskWise> jsonTaskWise = JsonUtil.readJsonFile(
            filePath, JsonSerializableTaskWise.class);

    if (jsonTaskWise.isEmpty()) {
        return Optional.empty();
    }

    try {
        return Optional.of(jsonTaskWise.get().toModelType());
    } catch (IllegalJsonValueException ive) {
        logger.info("Illegal values found in " + filePath + ": " + ive.getMessage());
        throw new FileStorageLoadException(ive);
    }
}
```

into 

```java
public ExOptional<ReadOnlyTaskWise> readTaskWise(Path filePath) throws FileStorageLoadException {
    requireNonNull(filePath);
    
    try {
        return JsonUtil.readJsonFile(filePath, JsonSerializableTaskWise.class)
                .filter(x -> !x.isEmpty())
                .map(x -> x.toModelType());
    } catch (IllegalJsonValueException ive) {
        logger.info("Illegal values found in " + filePath + ": " + ive.getMessage());
        throw new FileStorageLoadException(ive);
    }
}
```

In the above code, the method `readJsonFile` returns an `ExOptional`, which can then be mapped with an anonymous `ExFunction` object `x -> x.toModelType()`, which throws `IllegalJsonValueException`.

The inspiration for these changes came from https://stackoverflow.com/questions/27644361/how-can-i-throw-checked-exceptions-from-inside-java-8-lambdas-streams, and the implementation was inspired by my work in CS2030S.

**Other Minor Fixes to Code Quality**

I have also made minor improvements to the code quality of the codebase, by redoing some functional programming code snippets, as well as renaming some existing functions and classes to better suit the needs of our app.

### CODE REVIEW

**Review by JCSnap — COMMENTED** [review]
Looks legit! Maybe it can be helpful to include some potential use cases of ExOptionals and the ExFunctional and whatnot in the PR description

**Comment by asdfghjkxd** [comment]
@JCSnap I have included a short snippet of how the new `ExOptional` and `ExFunction` is used to simplify code and enhance existing functional programming code in the code base.

**Review by JCSnap — APPROVED** [review]

**Review by JCSnap — APPROVED** [review]

**Review by maypfv — COMMENTED** [review]
Looks good to me!

File: src/main/java/seedu/address/commons/core/functional/ExFunction.java
Comment (Sebtey) [inline src/main/java/seedu/address/commons/core/functional/ExFunction.java:1]: Hmm, how would this class be used?

**Review by seantehds — APPROVED** [review]
LGTM George! Nice one!

**Review by Sebtey — COMMENTED** [review]
LGTM, except that I don't quite get the purpose of the ExFunction file

**Pull request merged**